scripts/rddepman.ts

```diff
@@ -0,0 +1,197 @@
+  const title = getTitle(name, currentVersion, latestVersion);
+  const branchName = getBranchName(name);
+
+  console.log(`Creating PR "${ title }".`);
+  await getOctokit().rest.pulls.create({
+    owner: GITHUB_OWNER,
+    repo:  GITHUB_REPO,
+    title,
+    base:  MAIN_BRANCH,
+    head:  branchName,
+  });
+}
+
+async function getPulls(...options: Parameters<Octokit['rest']['pulls']['list']>): Promise<Awaited<ReturnType<Octokit['rest']['pulls']['list']>>['data']> {
+  let response: Awaited<ReturnType<Octokit['rest']['pulls']['list']>>;
+  let retries = 0;
+
+  while (true) {
+    try {
+      response = await getOctokit().rest.pulls.list(...options);
+      break;
+    } catch (error: any) {
+      retries += 1;
+      if (retries > 2) {
+        throw error;
+      }
+    }
+  }
+
+  return response.data;
+}
+
+async function checkDependencies(): Promise<void> {
+  // exit if there are unstaged changes
+  git('update-index', '--refresh');
+  if (git('diff-index', '--quiet', 'HEAD', '--')) {
+    console.log('You have unstaged changes. Commit or stash them to manage dependencies.');
+
+    return;
+  }
+
+  // load current versions of dependencies
+  const depVersionsPath = path.join('src', 'assets', 'dependencies.yaml');
+  const currentVersions = await readDependencyVersions(depVersionsPath);
+
+  // get a list of dependencies' version comparisons
+  const versionComparisons: VersionComparison[] = await Promise.all(dependencies.map(async(dependency) => {
+    return {
+      name:           dependency.name,
+      currentVersion: currentVersions[dependency.name as keyof DependencyVersions],
+      latestVersion:  await dependency.getLatestVersion(),
+    };
+  }));
+
+  // filter comparisons down to the ones that need an update
+  const updatesAvailable = versionComparisons.filter(({ name, currentVersion, latestVersion }) => {
+    const equal = compareVersions(currentVersion, latestVersion);
+
+    if (equal) {
+      console.log(`${ name } is up to date.`);
+    } else {
+      console.log(`Can update ${ name } from ${ currentVersion } to ${ latestVersion }`);
+    }
+
+    return !equal;
+  });
+
+  // reconcile dependencies that need an update with state of repo's PRs
+  const needToCreatePR: VersionComparison[] = [];
+
+  await Promise.all(updatesAvailable.map(async({ name, currentVersion, latestVersion }) => {
+    // try to find PR for this combo of name, current version and latest version
+    const branchName = getBranchName(name);
+
+    const prs = await getPulls({
+      owner: GITHUB_OWNER, repo: GITHUB_REPO, head: `${ GITHUB_OWNER }:${ branchName }`, state: 'all', per_page: 100,
+    });
+
+    const title = getTitle(name, currentVersion, latestVersion);
+    let prExists = false;
+
+    await Promise.all(prs.map(async(pr) => {
+      if (pr.title !== title && pr.state === 'open') {
+        console.log(`Closing stale PR "${ pr.title }".`);
```

Comment:
Alternatively, we could update the existing PR instead. Though I guess this would be better in case somebody had made manual commits on top, so they can reopen instead of having to sort out the tangle…

So, keep as it is I guess?

Reply:
Yeah. But your comment, and time, made me realize that there is a problem here. The way it is right now, all version bumps for a dependency go into a branch called `rddepman/<dependencyName>`. Every time there is a new version of a dependency, this branch is force pushed. So any commits made on top of that branch get nuked. Not the biggest problem, since a dev is likely to have a local copy of these commits, but I can see this being very annoying, and in some cases the cause of lost work.

I wrote it this way because of limited functionality in Github's PR API (can't search by partial branch name). But I just realized that I can use the search part of their API, which allows us to make this much better. I will do this shortly.